### PR TITLE
afaryy/TDL09

### PR DESCRIPTION
**Pull request by afaryy**
Add code for to-do-list-09.

### CODE REVIEW

File: TDL-09-shell-EC2/deploy_instance.sh [hunk 1]
```diff
@@ -0,0 +1,229 @@
...
+function usage(){
+  case "$1" in
+    all)
+      print_color "red" "To deploy your instance, please run deploy_instance.sh as follows:"
+      print_color "red" "./deploy_instance.sh ostype agedtime [instance_type]"
+      print_color "red" "The value range of the arguments: "
+      print_color "red" "1. valid ostype: amzn, amzn2, centos6, centos7, ubuntu"
+      print_color "red" "2. agedtime: 1-7 "
+      print_color "red" "3. instance_type: default t2_micro, valid instance_type: t2.micro, t2.small, t3.micro"
+      ;;
```
Comment (hbwork) [inline TDL-09-shell-EC2/deploy_instance.sh:89]: color red normally is used for error msg:)

File: TDL-09-shell-EC2/deploy_instance.sh [hunk 2]
```diff
@@ -0,0 +1,229 @@
...
+  echo $result
+  return $result
+}
+
+#######################################
+# Show how to use the deploy_instance.sh.
+# Arguments:
+#   usage_type. eg: all, ostype, agedtime, instance_type.
+#######################################
+function usage(){
```
Comment (hbwork) [inline TDL-09-shell-EC2/deploy_instance.sh:80]: Well-done:)

A little-bit over-engineered usage() function,  the simple way is to print out the usage information only with no parameter.

File: TDL-09-shell-EC2/deploy_instance.sh [hunk 3]
```diff
@@ -0,0 +1,229 @@
...
+if [[ $# -lt 2 || $# -gt 3 ]];  then
+  usage 'all'
+  exit
+fi
+
+# check whehter ostype is a valid option, if ostype is an invalid ostype, print usage() and exit 
+list_include_item "${LIST_OSTYPE[*]}" "$1"
+result=$?
+if (( result == 1 )); then
+  usage 'ostype'
```
Comment (hbwork) [inline TDL-09-shell-EC2/deploy_instance.sh:120]: if (( ${result} == 1 )) ?

File: TDL-09-shell-EC2/deploy_instance.sh [hunk 4]
```diff
@@ -0,0 +1,229 @@
...
+  exit
+fi
+ostype=$1
+
+# check whether 1 <= agedtime <=7, otherwise iprint usage() and exit
+if (( $2 <1 || $2 > 7)); then
```
Comment (hbwork) [inline TDL-09-shell-EC2/deploy_instance.sh]: space after <:) 

if (( ${2} < 1 || ${2} > 7)); then

File: TDL-09-shell-EC2/deploy_instance.sh [hunk 5]
```diff
@@ -0,0 +1,229 @@
...
+  usage 'agedtime'
+  exit
+fi
+agedtime=$2
+
+# check whether instance_type is valid, otherwise iprint usage() and exit
+if [[ $# -eq 3 ]];  then
```
Comment (hbwork) [inline TDL-09-shell-EC2/deploy_instance.sh:133]: instance_type=${3:-t2.micro} ?

which will set up the default value for instance_type

File: TDL-09-shell-EC2/deploy_instance.sh [hunk 6]
```diff
@@ -0,0 +1,229 @@
...
+    ;;
+esac
+echo "Find the latest image of $ostype: ${image_id}"
+ 
+#  Create EC2 instance for osype with instance_type with below settings:
+print_color "green" "-------------------------Create EC2 instance---------------------------"
+# set variables
+KEY_NAME='da-key'
+VPC_NAME='da-c02-vpc'
+VPC_ID=$(aws ec2 describe-vpcs \
```
Comment (hbwork) [inline TDL-09-shell-EC2/deploy_instance.sh]: It is better to define these variables at the very beginning of the script so you can update the code later easily in the future.

File: TDL-09-shell-EC2/deploy_instance.sh [hunk 7]
```diff
@@ -0,0 +1,229 @@
...
+# echo ${start_time}
+## AgedTime: yyyymmddhhmm (current time + agedtime days in Australia/Melbourne Time)
+stop_time=$(date -d "$(TZ=$DEFAULT_TIME_ZONE date) $agedtime days -3600 seconds" +%Y%m%d%H%M%S)
+echo ${stop_time}
+
+# Create user_data file
+cat << EOF > ./instance_userdata.txt
+#!/bin/bash
+cat <<EOT >> /etc/motd
+'OSType: $ostype'
```
Comment (hbwork) [inline TDL-09-shell-EC2/deploy_instance.sh]: 'OSType: ${ostype}'

File: TDL-09-shell-EC2/deploy_instance.sh [hunk 8]
```diff
@@ -0,0 +1,229 @@
...
+"Please be notified that your instance will be terminated by  ${stop_time}"
+EOT
+EOF
+
+EC2_NAME='ostype_test'
```
Comment (hbwork) [inline TDL-09-shell-EC2/deploy_instance.sh]: EC2_NAME="${ostype}_test"

File: TDL-09-shell-EC2/deploy_instance.sh [hunk 9]
```diff
@@ -0,0 +1,229 @@
...
+RULE='ec2-startstop;1900;;all'
```
Comment (hbwork) [inline TDL-09-shell-EC2/deploy_instance.sh]: RULE=';1900;;all'

File: TDL-09-shell-EC2/deploy_instance.sh [hunk 10]
```diff
@@ -0,0 +1,229 @@
...
+  echo $state
+  if [[ $state == 'running' ]]; then
+    print_color "green" "Successed. Instance is running."
+    break
+  fi
+done
+
+# Print out private ip and 
+print_color "green" "-------------------------print out instance id and private ip--------------------------"
+print_color "green" "Outputs: InstanceId=\"$instance_id\", PrivateIpAddress=\"$private_ip_address\""
```
Comment (hbwork) [inline TDL-09-shell-EC2/deploy_instance.sh]: One more thing: clean up, delete the temp file for userdata script;)

File: TDL-09-shell-EC2/deploy_instance.sh [hunk 11]
```diff
@@ -0,0 +1,229 @@
...
+instance_id=$(aws ec2 run-instances \
+    --image-id "$image_id" \
+    --instance-type "$instance_type" \
+    --count 1 \
+    --subnet-id "$PRIVATE_SUBNET_A_ID" \
+    --key-name  "$KEY_NAME" \
+    --security-group-ids "$SG_PRIVATE_ID" \
+    --no-associate-public-ip-address \
+    --user-data file://instance_userdata.txt  \
+    --tag-specifications "ResourceType=instance,Tags=[{Key=Name,Value=$EC2_NAME},{Key=AgedTime,Value=$stop_time},{Key=scheduler,Value=$RULE}]"  \
```
Comment (hbwork) [inline TDL-09-shell-EC2/deploy_instance.sh]: {Key=scheduler:ec2-startstop,Value=$RULE}

File: TDL-09-shell-EC2/deploy_instance.sh (hunk 1 above)
Comment (afaryy) [reply TDL-09-shell-EC2/deploy_instance.sh:89]: ok. Thanks.

File: TDL-09-shell-EC2/deploy_instance.sh (hunk 2 above)
Comment (afaryy) [reply TDL-09-shell-EC2/deploy_instance.sh:80]: :)

File: TDL-09-shell-EC2/deploy_instance.sh (hunk 3 above)
Comment (afaryy) [reply TDL-09-shell-EC2/deploy_instance.sh:120]: in (()) ${} is unneccessary.

File: TDL-09-shell-EC2/deploy_instance.sh (hunk 4 above)
Comment (afaryy) [reply TDL-09-shell-EC2/deploy_instance.sh]: haha,ok.

File: TDL-09-shell-EC2/deploy_instance.sh (hunk 5 above)
Comment (afaryy) [reply TDL-09-shell-EC2/deploy_instance.sh:133]: Good idea. Thanks

File: TDL-09-shell-EC2/deploy_instance.sh (hunk 6 above)
Comment (afaryy) [reply TDL-09-shell-EC2/deploy_instance.sh]: Ok.

File: TDL-09-shell-EC2/deploy_instance.sh (hunk 7 above)
Comment (afaryy) [reply TDL-09-shell-EC2/deploy_instance.sh]: ok.thanks.

File: TDL-09-shell-EC2/deploy_instance.sh (hunk 8 above)
Comment (afaryy) [reply TDL-09-shell-EC2/deploy_instance.sh]: Thank you :)

File: TDL-09-shell-EC2/deploy_instance.sh (hunk 9 above)
Comment (afaryy) [reply TDL-09-shell-EC2/deploy_instance.sh]: :( yeah, a mistake.

File: TDL-09-shell-EC2/deploy_instance.sh (hunk 10 above)
Comment (afaryy) [reply TDL-09-shell-EC2/deploy_instance.sh]: Ok.

File: TDL-09-shell-EC2/deploy_instance.sh (hunk 11 above)
Comment (afaryy) [reply TDL-09-shell-EC2/deploy_instance.sh]: I have corrected it.

**Comment by afaryy** [comment]
I have fixed the to-do-list-09. Thanks.

**Pull request merged**